src/address.ts

```diff
@@ -259,3 +268,70 @@ export class Address {
         return this.toHex().startsWith(SMART_CONTRACT_HEX_PUBKEY_PREFIX);
     }
 }
+
+export class AddressComputer {
+    private readonly numberOfShardsWithoutMeta: number;
+
+    constructor(numberOfShardsWithoutMeta?: number) {
+        this.numberOfShardsWithoutMeta = numberOfShardsWithoutMeta || CURRENT_NUMBER_OF_SHARDS_WITHOUT_META;
+    }
+
+    computeContractAddress(deployer: IAddress, deploymentNonce: bigint): Address {
+        let initialPadding = Buffer.alloc(8, 0);
+        let ownerPubkey = deployer.getPublicKey();
+        let shardSelector = ownerPubkey.slice(30);
+        let ownerNonceBytes = Buffer.alloc(8);
+
+        const bigNonce = new BigNumber(deploymentNonce.toString());
+        const bigNonceBuffer = bigIntToBuffer(bigNonce);
+        ownerNonceBytes.write(bigNonceBuffer.reverse().toString("hex"), "hex");
+
+        let bytesToHash = Buffer.concat([ownerPubkey, ownerNonceBytes]);
+        let hash = createKeccakHash("keccak256").update(bytesToHash).digest();
+        let vmTypeBytes = Buffer.from(WasmVirtualMachine, "hex");
+        let addressBytes = Buffer.concat([initialPadding, vmTypeBytes, hash.slice(10, 30), shardSelector]);
+
+        let address = new Address(addressBytes);
+        return address;
+    }
+
+    getShardOfAddress(address: IAddress): number {
+        return this.getShardOfPubkey(address.getPublicKey(), this.numberOfShardsWithoutMeta);
+    }
+
+    private getShardOfPubkey(pubkey: Uint8Array, numberOfShards: number): number {
```

Comment:
For review, linking to the Go and PY implementations:
 - https://github.com/multiversx/mx-chain-core-go/blob/main/core/sharding/multiShardCoordinator.go#L52
 - https://github.com/multiversx/mx-sdk-py-core/blob/main/multiversx_sdk_core/address.py#L141

This works for 1, 2, 3 and 4 shards. But will fail for more. See the Go implementation. For the moment, we can keep the PY and JS implementations as they are, but we should fix them in the future.

Reply:
Will replicate the Go implementation in the future.